src/jit/compiler.h

```diff
@@ -2088,6 +2107,7 @@ class Compiler
     GenTreeArgList* gtNewArgList(GenTreePtr op);
     GenTreeArgList* gtNewArgList(GenTreePtr op1, GenTreePtr op2);
     GenTreeArgList* gtNewArgList(GenTreePtr op1, GenTreePtr op2, GenTreePtr op3);
+    GenTreeArgList* gtNewArgList(GenTreePtr op1, GenTreePtr op2, GenTreePtr op3, GenTreePtr op4);
```

Comment:
In the SSE2, AVX and AVX-512 there are 16, 32, and 64 arguments in set_epi8 intrinsic series. What would be the best way to handle them? Creating single GenTreeArgList? Any ideas? I am working now on 16 byte SSE2 set_epi8 and other overloads and it seems there is a need for general design to handle it efficiently.

Reply:
I think a `GenTreeArgList` is probably the easiest way to handle it today.

Given the number of arguments (and infrequency of intrinics that take more than 4 operands), it is probably best to dynamically create the ArgList using a loop, rather than creating custom constructors for them.

Reply:
We can using something like:
```cpp
GenTree* Compiler::buildArgList(int num, var_types ty)
{
    assert(num >= 2);
    GenTree* list = nullptr;
    for (int i = 0; i < num; i++)
    {
        GenTree* arg = impSIMDPopStack(ty);
        list         = new (this, GT_LIST) GenTreeOp(GT_LIST, ty, arg, list);
    }
    return list;
}
```

Reply:
Thanks! How about using an array of GenTreePtr, we recover them anyway while calling impPopStack()?

Reply:
> How about using an array of GenTreePtr, we recover them anyway while calling 

There are intermediate steps between the creation of the HWIntrinsic node and the codegen phase that would need to be updated to support that, and it may be more trouble than its worth (I had looked at updating some of this so I could have gtOp1 be the first operand and gtOp2 be an ArgList for the remaining, and even that was non-trivial).

> We can have something like:
> `GenTree* Compiler::buildArgList(int num, var_types ty)`

Given that most of the higher argument `Set` intrinsics take a multiple of 4 arguments, we might be able to optimize that a bit more and turn it into a chain of `gtNewArgList(op1, op2, op3, op4)` calls.

Reply:
@4creators I am polishing my table-driven code that already contains this helper function, and I think it can be submitted by the end of tomorrow. Once we have a final solution to the table-driven approach, these infrastructure codes may need to change. I do not want to waste your and @tannergooding's valuable work and time (sorry to block your work), so I recommend to suspend this work for a little while.

Reply:
> and it may be more trouble than its worth

If that's the case let's use `GenTreeArgList`.

Reply:
> Given that most of the higher argument Set intrinsics take a multiple of 4 arguments, we might be able to optimize that a bit more and turn it into a chain of gtNewArgList(op1, op2, op3, op4) calls.

`Set` is special, as previously mentioned, these helper-intrinsics can be transformed in the importer, so that I do not need a long arg list actually.

Reply:
Ah, yes. That's on the list of things I want to fix for this PR 😄

Reply:
OK - should I in this case delay implementing `Set` until we have transformation support in the importer or go ahead and later refactor?

Reply:
The advantage of implementing now would be that we can get earlier to complete Sse2 ISA implementation.

Reply:
@4creators, `impSSE2Intrinsic` is part of the importer, so you should just be able to construct and return the appropriate node chain there.

Ex: The following "should" work (currently compiling to validate):
```
case NI_SSE_Set:
{
    assert(sig->numArgs == 4);
    assert(getBaseTypeOfSIMDType(sig->retTypeSigClass) == TYP_FLOAT);

    op4 = impPopStack().val;
    op3 = impPopStack().val;
    op2 = impPopStack().val;
    op1 = impPopStack().val;

    retNode = gtNewSimdHWIntrinsicNode(TYP_SIMD16, 
        gtNewSimdHWIntrinsicNode(TYP_SIMD16, op4, op3, NI_SSE_UnpackLow, TYP_FLOAT, 16),
        gtNewSimdHWIntrinsicNode(TYP_SIMD16, op2, op1, NI_SSE_UnpackLow, TYP_FLOAT, 16),
        gtNewIconNode(68, TYP_UBYTE),
        NI_SSE_Shuffle, TYP_FLOAT, 16
    );
    break;
}
```

Reply:
> impSSE2Intrinsic is part of the importer, so you should just be able to construct and return the appropriate node chain there

OK than it seems that some experiments should not hurt

Reply:
> The advantage of implementing now would be that we can get earlier to complete Sse2 ISA implementation.

Maybe not, we must unify the implementations of all the ISA classes. My plan is
1. complete table-driven infrastructure,
2. finish all the simple intrinsic in AVX/AVX2/SSE/SSE2/SSE41, so I recommend you guys to split the huge PRs.
3. move to complicated guys (e.g., memory access and IMM intrinsics).
4. ...

During this process, if anything blocks your work, we can move to https://github.com/dotnet/coreclr/issues/15490 that looks more independent.